### ISSUE TITLE
Proper error msg in case of issue with plan file

### ISSUE THREAD

**Issue by raj-dhamsaniya** [infracost/infracost#2139]
What is the issue?
- In one of the instances, in the tfplan json file, there are quotes(`''`) added at the start and end of the file. Because of this, infracost failed with providing help in running the diff command. however, the actual issue was with the plan file. Here is the debug output of the same. 

> infracost diff --path plan.json --log-level debug
time="2022-11-17T16:15:59+05:30" level=debug currency=USD func="github.com/infracost/infracost/internal/config.(*RunContext).IsCloudEnabled" file="github.com/infracost/infracost/internal/config/run_context.go:192" sync_usage=false enable_cloud_org=false is_cloud_enabled=false msg="IsCloudEnabled inferred from Config.EnabledDashboard"
time="2022-11-17T16:15:59+05:30" level=debug enable_cloud_org=false func="github.com/infracost/infracost/internal/apiclient.(*APIClient).doRequest" file="github.com/infracost/infracost/internal/apiclient/client.go:58" currency=USD sync_usage=false msg="'POST' request to '/graphql' using trace_id: 'db77678e-f1af-46e6-ad20-ad15280cd5fe'"
time="2022-11-17T16:16:00+05:30" level=debug result="{CloudEnabled:false}" func=main.loadCloudSettings file="github.com/infracost/infracost/cmd/infracost/main.go:251" currency=USD sync_usage=false enable_cloud_org=false msg="Successfully loaded settings from Infracost Cloud"
Show diff of monthly costs between current and planned state
> 
> USAGE
  infracost diff [flags]
> 
> EXAMPLES
  Use Terraform directory:
>
>     infracost breakdown --path /code --format json --out-file infracost-base.json
>     # Make Terraform code changes
>     infracost diff --path /code --compare-to infracost-base.json
>
>  Use Terraform plan JSON:
>
>      terraform plan -out tfplan.binary
>      terraform show -json tfplan.binary > plan.json
>      infracost diff --path plan.json
>
> FLAGS
>      --compare-to string            Path to Infracost JSON file to compare against
>      --config-file string           Path to Infracost config file. Cannot be used with path, terraform* or usage-file flags
>      --exclude-path strings         Paths of directories to exclude, glob patterns need quotes
>      --format string                Output format: json, diff (default "diff")
>  -h, --help                         help for diff
>      --no-cache                     Don't attempt to cache Terraform plans
>      --out-file string              Save output to a file
>  -p, --path string                  Path to the Terraform directory or JSON/plan file
>      --project-name string          Name of project in the output. Defaults to path or git repo name
>      --show-skipped                 List unsupported and free resources
>      --sync-usage-file              Sync usage-file with missing resources, needs usage-file too (experimental)
>      --terraform-var strings        Set value for an input variable, similar to Terraform's -var flag
>      --terraform-var-file strings   Load variable files, similar to Terraform's -var-file flag. Provided files must be relative to the >--path flag
>      --terraform-workspace string   Terraform workspace to use. Applicable when path is a Terraform directory
>      --usage-file string            Path to Infracost usage file that specifies values for usage-based resources
>
>GLOBAL FLAGS
>      --debug-report       Generate a debug report file which can be sent to Infracost team
>      --log-level string   Log level (trace, debug, info, warn, error, fatal)
>      --no-color           Turn off colored output
>
>Error: To show a diff:
>  1. Generate a cost estimate baseline: `infracost breakdown --path plan.json --format json --out-file infracost-base.json`
>  2. Make a Terraform code change
>  3. Generate a cost estimate diff: `infracost diff --path plan.json --compare-to infracost-base.json`
time="2022-11-17T16:16:00+05:30" level=debug enable_cloud_org=false func="github.com/infracost/infracost/internal/apiclient.(*APIClient).doRequest" file="github.com/infracost/infracost/internal/apiclient/client.go:58" currency=USD sync_usage=false msg="'POST' request to '/event' using trace_id: 'db77678e-f1af-46e6-ad20-ad15280cd5fe'"

Why its require?
- To avoid confusion and better error handling

What can be the alternative output?
- Instead of this we can provide the info that the plan is not parsable and what all version/plan is supported.

Suggestion for improvements
- We can check by parsing the file that a json is able to parse, if not throw this error.

**Comment by vdmgolub** [comment]
@raj-dhamsaniya Hello! Thank you for creating the issue!

The error you're seeing is not caused by the invalid tfplan JSON file but by the missing required `--compare-to` flag. The infracost diff` command compares your current plan with some baseline. Hence it shows you instructions with 3 steps: 1. generate baseline, 2. make changes, 3. run diff to see the cost difference made by your changes.

If you provide the flag, you should receive an error `Error: No valid Terraform files found at the given path, try a different directory` indicating that your file has an issue.

Hope this explains it. Please let me know if you have more questions :)

**Comment by vdmgolub** [comment]
Yep, my bad. Sorry for the confusion. I forgot that with the TF plan, we could detect both states. I was able to reproduce it as well.

The `--path` flag supports different options, and we already try to parse JSON to detect if it's a valid JSON file. When this check fails, we keep checking other possible options, and if none are detected, we default to the Terraform directory. Thus the output message. So the improvement you're suggesting is already in place :)

However, as you said, we could improve the messaging as it currently shows instructions for the TF directory option. We could include something like "If you're using Terraform plan, please verify it is a valid JSON file". I'll bring this to the team to discuss what is a better messaging. Thanks again for reporting the issue :)

**Comment by raj-dhamsaniya** [comment]
Was looking for the code. thanks for this. I see why the issue is there. makes sense. 

IMHO we can have two options i.e. path for terraform dir approach or --plan-path for terraform plan. 
Saying that I do see the intuition behind keeping it the same flag but the way we want to derive the diff is entirely diff so the diff flags do make sense.

**Comment by vdmgolub** [comment]
It is a constant battle between supporting more options and keeping the number of flags at bay. We went through several iterations to make the DX better. Nothing is ideal. There are always ways to improve. Thanks to you and the community, we can spot and fix that. :D

**Comment by glenngillen** [comment]
We've reworked all of the our CLI and will be releasing an updated CLI in the coming days. Closing this issue as it won't be relevant to the new one.
